fs/backgroundfetcher/background_fetcher.go

```diff
@@ -0,0 +1,152 @@
+)
+
+type Option func(*BackgroundFetcher) error
+
+const (
+	// Amount of time the background fetcher will wait once a new layer comes in
+	// before (re)starting fetches.
+	defaultSilencePeriod = 5 * time.Second
+
+	// The minimum size of data to be fetched in a single fetch.
+	defaultMinFetchSize = 4 * 1024 * 1024 // 4 MiB
+
+	// Specifies how often the fetch will occur.
+	// The background fetcher will fetch a single span every `defaultFetchPeriod`.
+	defaultFetchPeriod = 500 * time.Millisecond
+)
+
+func WithMinFetchSize(size int64) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.minFetchSize = size
+		return nil
+	}
+}
+
+func WithSilencePeriod(period time.Duration) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.silencePeriod = period
+		return nil
+	}
+}
+
+func WithFetchPeriod(period time.Duration) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.fetchPeriod = period
+		return nil
+	}
+}
+
+// A backgroundFetcher is responsible for fetching spans from layers
+// in the background.
+type BackgroundFetcher struct {
+	minFetchSize  int64
+	silencePeriod time.Duration
+	fetchPeriod   time.Duration
+
+	rateLimiter *rate.Limiter
+
+	// All span managers are added to the channel and picked up in Run().
+	// If a span manager is still able to fetch, it is reinserted into the chanel.
+	waitQueue chan Resolver
+
+	closeChan chan struct{}
+
+	// The timer that the bg-fetcher will wait on until it is ready to (re)start
+	// background fetches.
+	timer *time.Timer
+}
+
+func NewBackgroundFetcher(opts ...Option) (*BackgroundFetcher, error) {
+	bf := newWithDefaults()
+	for _, o := range opts {
+		if err := o(bf); err != nil {
+			return nil, err
+		}
+	}
+	// Create a rate-limiter that will fetch every bf.fetchPeriod
+	// with a burst capacity of 1 (i.e., it will never invoke more than 1 bg-fetch
+	// within bf.fetchPeriod)
+	bf.rateLimiter = rate.NewLimiter(rate.Every(bf.fetchPeriod), 1)
+	return bf, nil
+}
+
+func newWithDefaults() *BackgroundFetcher {
+	return &BackgroundFetcher{
+		minFetchSize:  defaultMinFetchSize,
+		silencePeriod: defaultSilencePeriod,
+		fetchPeriod:   defaultFetchPeriod,
+		waitQueue:     make(chan Resolver),
+		closeChan:     make(chan struct{}),
+	}
+}
+
+// Add a new Resolver to be background fetched from.
+// Sends the resolver through the channel, which will be received in the Run() method.
+// Also resets the `timeTicker` so that background fetches are suspended until `silencePeriod`.
+func (bf *BackgroundFetcher) Add(resolver Resolver) {
+	bf.waitQueue <- resolver
+	if bf.timer == nil {
+		bf.timer = time.NewTimer(bf.silencePeriod)
+	} else {
+		bf.timer.Reset(bf.silencePeriod)
+	}
+}
+
+func (bf *BackgroundFetcher) Close() error {
+	bf.closeChan <- struct{}{}
+	return nil
+}
+
+func (bf *BackgroundFetcher) Run(ctx context.Context) error {
+	for {
+
+		// If any new span managers have come in, wait for silencePeriod.
+		if bf.timer != nil {
+			<-bf.timer.C
+		}
```

Comment:
Is it possible to put this channel read in the select below? Ideally speaking, the passed context's cancellation should be honored while waiting this timer.

Reply:
Done.